source/common/upstream/cluster_manager_impl.cc

```diff
@@ -617,10 +617,14 @@ bool ClusterManagerImpl::addOrUpdateCluster(const envoy::config::cluster::v3::Cl
   const auto existing_active_cluster = active_clusters_.find(cluster_name);
   const auto existing_warming_cluster = warming_clusters_.find(cluster_name);
   const uint64_t new_hash = MessageUtil::hash(cluster);
-  if ((existing_active_cluster != active_clusters_.end() &&
-       existing_active_cluster->second->blockUpdate(new_hash)) ||
-      (existing_warming_cluster != warming_clusters_.end() &&
-       existing_warming_cluster->second->blockUpdate(new_hash))) {
+  if (existing_warming_cluster != warming_clusters_.end()) {
+    if (existing_warming_cluster->second->blockUpdate(new_hash)) {
```

Comment:
Can you generally add more comments to this entire sequence? It's honestly pretty confusing at this point to understand all the branches. Thank you!